airbyte-integrations/connectors/source-sendgrid/source_sendgrid/api.py

```diff
@@ -0,0 +1,202 @@
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        params = {"limit": self.limit}
+        if next_page_token:
+            params.update(**next_page_token)
+        return params
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        stream_data = response.json()
+        if self.data_field:
+            stream_data = stream_data[self.data_field]
+        if len(stream_data) < self.limit:
+            return
+        self.offset += self.limit
+        return {"offset": self.offset}
+
+
+class SendgridStreamMetadataPagination(SendgridStream):
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        params = {}
+        if not next_page_token:
+            params = {"page_size": self.limit}
+        return params
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        next_page_url = response.json()["_metadata"].get("next", False)
+        if next_page_url:
+            return {"next_page_url": next_page_url.replace(self.url_base, "")}
+
+    @staticmethod
+    @abstractmethod
+    def initial_path() -> str:
+        """
+        :return: initial path for the API endpoint if no next metadata url found
+        """
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        if next_page_token:
+            return next_page_token["next_page_url"]
+        return self.initial_path()
+
+
+class Scopes(SendgridStream):
+    def path(self, **kwargs) -> str:
+        return "scopes"
+
+
+class Lists(SendgridStreamMetadataPagination):
+    data_field = "result"
+
+    @staticmethod
+    def initial_path() -> str:
+        return "marketing/lists"
+
+
+class Campaigns(SendgridStreamMetadataPagination):
+    data_field = "result"
+
+    @staticmethod
+    def initial_path() -> str:
+        return "marketing/campaigns"
+
+
+class Contacts(SendgridStream):
+    data_field = "result"
+
+    def path(self, **kwargs) -> str:
+        return "marketing/contacts"
+
+
+class StatsAutomations(SendgridStreamMetadataPagination):
+    data_field = "results"
+
+    @staticmethod
+    def initial_path() -> str:
+        return "marketing/stats/automations"
+
+
+class Segments(SendgridStream):
+    data_field = "results"
+
+    def path(self, **kwargs) -> str:
+        return "marketing/segments"
+
+
+class Templates(SendgridStreamMetadataPagination):
+    data_field = "result"
+
+    def request_params(self, next_page_token: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(next_page_token=next_page_token, **kwargs)
+        if not next_page_token:
```

Comment:
why do we selectively set generations based on next page token?

Reply:
I believe because the next_page_token is a complete URL with parameters, so don't want to mention them twice (could be treated like an array? dunno)